Add HLSL vector/matrix template declaration option

pow2clk opened the pull request:
Refactors vector and matrix docs pages to allow more self-referencing of concepts, clear up some confusing parts, adds examples, and finally adds the default values of the template parameters that allows them to be excluded to get the defaults, which was the actual motivator of this change.

Fixes https://github.com/microsoft/DirectXShaderCompiler/issues/2034

Commit:
Add HLSL vector/matrix template declaration option

Refactors vector and matrix docs pages to allow more
self-referencing of concepts as well as clear up some unclear
parts and adds the default values of the template parameters that
allows them to be excluded to get the defaults.

python3kgae commented on desktop-src/direct3dhlsl/dx-graphics-hlsl-vector.md:
````diff
@@ -16,80 +16,47 @@ ms.date: 05/31/2018
-
-```
+```hlsl
 bool    bVector;   // scalar containing 1 Boolean
````
Should this be bool1 instead of bool?

pow2clk replied:
It's pre-existing. I think the intent was to demonstrate the difference between declaring something as a scalar versus as a vector. I don't think that's terribly useful myself, but I'm ambivalent.

pow2clk replied:
I've revised the names and comments

Commit:
Revise boolean scalar example

The boolean scalar was, at best, confusingly named since it was called
`bVector`. I've changed it to an integer scalar with an appropriate name
to contrast with the next example of a single-element integer vector

siliconvoodoo commented on desktop-src/direct3dhlsl/dx-graphics-hlsl-matrix.md:
````diff
@@ -17,83 +17,62 @@ api_location:
 
-
-```
-matrix <float, 2, 2> fMatrix = { 0.0f, 0.1, // row 1
-                                 2.1f, 2.2f // row 2
+```hlsl
+matrix <int, 1, 1>   iMatrix = { 1 }; 
+matrix <float, 2, 3> fMatrix = { 0.0f, 0.1f, 0.2f, // row 1
+                                 2.1f, 2.2f, 2.3f  // row 2
                                };
+matrix<int16_t, 1>   sMatrix = { 1, 2, 3, 4 }; // Defaults to 1x4 int16 matrix
+matrix<float16_t>    hMatrix = { 0.0f, 0.1f, 0.2f, 0.3f, // Defaults to 4x4 float16 matrix
+                                 1.0f, 1.1f, 1.2f, 1.3f,
+                                 2.0f, 2.1f, 2.2f, 2.3f,
+                                 3.0f, 3.1f, 3.2f, 3.3f }; 
+matrix               fMatrix = { 0.0f, 0.1f, 0.2f, 0.3f, // Defaults to 4x4 float matrix
+                                 1.0f, 1.1f, 1.2f, 1.3f,
+                                 2.0f, 2.1f, 2.2f, 2.3f,
+                                 3.0f, 3.1f, 3.2f, 3.3f }; 
````
nice
